crates/apollo-compiler/src/validation/directive.rs

```diff
@@ -66,7 +66,16 @@ impl FindRecursiveDirective<'_> {
 
         let type_name = input_value.ty.inner_named_type();
         if let Some(type_def) = self.schema.types.get(type_name) {
-            self.type_definition(seen, type_def)?;
+            if seen.contains(type_def.name()) {
```

Comment:
Now, the RecursionGuard holds both directive names and type names. But, since the type namespace and directive namespace are separate, collision is possible. We can mangle names to distinguish them.
```suggestion
            let mangled_type_name = Name::new_unchecked(&format!("!type_name!{}", type_def.name()));
            if seen.contains(&mangled_type_name) {
```

Reply:
I think I'd prefer two separate `RecursionGuard`s over name mangling